### ISSUE TITLE
feature request: automatic conversion of numpy to torch tensors

### ISSUE THREAD

**Issue by kirk86** [pytorch/pytorch#1018]
Hi guys, I was wondering if it could be possible to have automatic conversion of numpy arrays implicitly when we use `model.forward(x)`where `x` is a numpy array and let torch implicitly transform it to torch tensors. Or when we use it to train the model. That would increase the time of implementation on the user side.

**Comment by soumith** [comment]
this is an explicit choice we made, to not auto-convert numpy arrays. Because the magical wrapping would essentially do: numpy->Tensor->Variable, a 3 hop magic.

I'll discuss it again with the others but I am not sure our decision will change.

Also, `model.forward(x)` is not correct, you need to do: `model(x)`. 

Otherwise hooks wont work.

Thanks for the suggestion!

**Comment by kirk86** [comment]
@soumith Thanks for your answer. Am I missing sth. In the simple convnet example using mnist. There's `model = Net()` so after training to evaluate my model shouldn't I do `model.forward(x)` or is it as you've said `model(x)`  to evaluate the model? I'm a bit confused here.

**Comment by soumith** [comment]
always do `model(x)`. `model.forward(x)` shouldn't be used, because hooks are not called in that case. If there is a place in the tutorials you picked this from, can you point me to it, I will change it immediately.

**Comment by kirk86** [comment]
@soumith I'll have a look again more closely and let you know. Since I'm coming from torch I intuitively might have thought that `model.forward(x)` might have been the natural thing to do. On a second note now I'm thinking that this might explain the different results that I've got using pytorch against keras on the same model with the exact same data. Pytorch got me ~94.x% while keras gave me ~97.x%

**Comment by soumith** [comment]
@kirk86 if you are not using pytorch's hooks feature (via `register_hook`) it probably doesn't explain the accuracy difference.

**Comment by kirk86** [comment]
@soumith no I'm not using `register_hook`. I suppose that is sth that you'd have to do explicitly right? The only thing that I'm using is simply an MLP model. I'll try to test again and if so I'll give an a simple snippet example replicating the issue. Thanks again for your help.
